src/controller/CHIPDeviceController.cpp

```diff
@@ -1178,6 +1184,298 @@ void DeviceCommissioner::OnSessionEstablished()
+
+    // Step g
+    VerifyOrReturnError(sizeof(mAttestationNonce) == attestationNonce.size(), CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(memcmp(mAttestationNonce, attestationNonce.data(), sizeof(mAttestationNonce)) == 0, CHIP_ERROR_INTERNAL);
+
+    // Step h/i
+    ReturnErrorOnFailure(ValidateCertificateDeclaration(certDeclaration, mRemoteManufacturerPubkey, firmwareInfo));
+
+    ChipLogProgress(Controller, "Sending Trusted Root Certificate to the device.");
+    ReturnErrorOnFailure(SendTrustedRootCertificate(device));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DeviceCommissioner::ValidateCertificateDeclaration(chip::ByteSpan certDeclaration, Crypto::P256PublicKey pubkey,
+                                                              chip::ByteSpan firmwareInfo)
+{
+    P256ECDSASignature signature;
+    ByteSpan certDeclarationTbs;
+    ByteSpan certDeclarationSignature;
+
+    const uint8_t * data = nullptr;
+    TLV::TLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+    tlvReader.Init(certDeclaration.data(), static_cast<uint32_t>(certDeclaration.size()));
+    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+    ReturnErrorOnFailure(tlvReader.Next());
+    ReturnErrorOnFailure(tlvReader.GetDataPtr(data));
+    certDeclarationTbs = ByteSpan(data, tlvReader.GetLength());
+    ReturnErrorOnFailure(tlvReader.Next());
+    VerifyOrReturnError(TLV::IsContextTag(tlvReader.GetTag()) == true, CHIP_ERROR_INTERNAL);
+    ReturnErrorOnFailure(tlvReader.GetDataPtr(data));
+    certDeclarationSignature = ByteSpan(data, tlvReader.GetLength());
+    ReturnErrorOnFailure(signature.SetLength(certDeclarationSignature.size()));
+    memcpy(signature, certDeclarationSignature.data(), certDeclarationSignature.size());
+
+    ReturnErrorOnFailure(pubkey.ECDSA_validate_msg_signature(certDeclarationTbs.data(), certDeclarationTbs.size(), signature));
+
+    uint16_t vendor_id;
+    uint16_t product_id;
+    uint16_t server_category_id;
+    uint16_t client_category_id;
+    uint16_t security_level;
+    uint16_t security_information;
+    uint16_t version_number;
+
+    uint16_t * element_array[] = { &vendor_id,      &product_id,           &server_category_id, &client_category_id,
+                                   &security_level, &security_information, &version_number };
+
+    int i         = 0;
+    containerType = TLV::kTLVType_Structure;
+    tlvReader.Init(certDeclarationTbs.data(), static_cast<uint32_t>(certDeclarationTbs.size()));
+    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+    do
+    {
+        if (!TLV::IsContextTag(tlvReader.GetTag()))
+        {
+            continue;
+        }
+        ReturnErrorOnFailure(tlvReader.Get(*element_array[i++]));
+    } while (tlvReader.Next() == CHIP_NO_ERROR);
```

Comment:
```suggestion
    uint16_t * element_array[] = { &vendor_id,      &product_id,           &server_category_id, &client_category_id,
                                   &security_level, &security_information, &version_number };

    int i         = 0;
    containerType = TLV::kTLVType_Structure;
    tlvReader.Init(certDeclarationTbs.data(), static_cast<uint32_t>(certDeclarationTbs.size()));
    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
    do
    {
        if (!TLV::IsContextTag(tlvReader.GetTag()))
        {
            continue;
        }
        ReturnErrorOnFailure(tlvReader.Get(*element_array[i++]));
    } while (tlvReader.Next() == CHIP_NO_ERROR);
```

This reading code again expects that every element is present in the correct order, **and only those**. Buffer overrun can occur. Please separate this parsing in a separate method and provide unit tests.

Reply:
fixed